Copter: high vibration failsafe

rmackay9 opened the pull request:
This PR adds a "high vibration failsafe" to Copter resolving these two issues ([vibration failsafe](https://github.com/ArduPilot/ardupilot/issues/4825), [get-vert-pos-rate returns zero](https://github.com/ArduPilot/ardupilot/issues/11119))

High vibration and loss of altitude control are detected if the following are true:

- EKF2's velocity and position innovations are positive
- EKF2's vertical velocity variation > 1

High vibrations can be tested in SITL by doing the artificially forcing IMU clipping by doing the following:

- param set EK2_IMU_MASK 1
- param set TUNE 58 (to allow ch6 to control the accelerometer range)
- param set TUNE_MAX 100
- param set SIM_ACC_RND 0
- arm in Loiter and takeoff to an altitude of 20m
- param set SIM_ACC_RND 60

At this point the vehicle will climb quickly as shown below:
![master-alt](https://user-images.githubusercontent.com/1498098/65812302-5bd3df00-e200-11e9-870e-fd85352dbbbc.png)

With this vibration failsafe the vehicle does not climb:
![vibe-failsafe-alt](https://user-images.githubusercontent.com/1498098/65812308-760dbd00-e200-11e9-8550-3ff84a7b96e5.png)

There are these known problems:

- [ ] Only implemented for the EKF2
- [x] if the MOT_THST_HOVER value is more than about 10% higher or lower than the vehicle's actual hover throttle the vehicle may be unable to descend (if the hover throttle is too high) or it may descend too quickly (if hover throttle is too low)

lukezhqin commented:
very interesting! I indeed met this condition in my own flying and havn't come up with a better idea than this.

amilcarlucas commented on ArduCopter/defines.h:
```diff
@@ -94,7 +94,8 @@ enum tuning_func {
     TUNING_RATE_YAW_FILT =              56, // yaw rate input filter
-    TUNING_WINCH =                      57  // winch control (not actually a value to be tuned)
+    TUNING_WINCH =                      57, // winch control (not actually a value to be tuned)
+    TUNING_ARTIFICIAL_CLIPPING =        58
```
Please add a comment, "accelerometer clipping for high vibration failsafe testing" or something like that

rmackay9 replied:
@amilcarlucas, thanks for having a look.  This is one of the commits with "-- do not merge" in the title though so this change won't make it into master... it's only to allow easier testing of this change.

priseborough commented on libraries/AP_NavEKF2/AP_NavEKF2_core.cpp:
```diff
@@ -758,6 +757,24 @@ void NavEKF2_core::calcOutputStates()
+
+    // Perform filter calculation using backwards Euler integration
+    // Coefficients selected to place all three filter poles at omega
+    const float CompFiltOmega = 10.0f;
```
This can be increased to use baro across wider bandwidth which reduces reliance on IMU and vice-versa

priseborough replied:
The filter integrates IMU to get vertical velocity and position and then corrects using baro, so the bigger this gain, the more the baro corrects.

rmackay9 replied:
I tried 3 different values for CompFiltOmega (10, 30 and 50) and in SITL at least I didn't really notice much difference in performance.  I have logs if that's of interest.

rmackay9 commented:
I've rebased on master and made these changes:

- made sure the vehicle part works with EKF3 (the EKF3::get_vert_pos_rate() function still needs fixing)
- added a new FS_VIBE_ENABLE parameter

I have not modified the landing detector to increase the acceptable accelerations because at least in SITL testing it's nearly impossible for it to detect the landing while there are any vibes.  SITL is not really realistic though because the vibrations don't decrease as the throttle comes down.  My point is perhaps we need some feedback from real users.

rmackay9 commented:
I test flew this today by forcing the vibration failsafe to trigger (even though it had low accelerations).  It oscillated a fair bit vertically especially if surface tracking was enabled.  The landing detector also never worked which means that it's not the high accelerations that are the problem but rather the noisy velocities from the EKF.

I'm considering trying to use the baro or GPS climb rate instead to see how much better or worse it is.

priseborough commented:
I will add a tuning parameter for the cross-over frequency

priseborough commented:
See https://github.com/priseborough/ardupilot/commit/a9ab6d45df07f0bf609869dabb481938c01789c8

Higher values for EK2_HRT_FILT should cause the height rate estimate to weight the Baro data more and the IMU data less. If you supply a log from boot with replay logging enabled, then I can try to tune the cross-over frequency using replay.

Unless we introduce a reduced bandwidth mode for the height controller, we will continue to be forced to incorporate some accelerometer data in the alternative height rate calculation which is not ideal when recovering from a situation caused by bad accel measurements.

priseborough commented:
I'll implement the same for EKF3 so it has a height rate available as per the interface. However i think it would still be worthwhile looking at the option of setting the accel process noise to a very large value while the clipping condition exists to force the estimates to follow the GPS vertical velocity and baro more closely.

rmackay9 commented:
@priseborough, thanks.  I've pulled that commit in from your branch.  This will be my top priority for this week so that we can get Copter-4.0.0 out for beta testing.

priseborough commented:
The height derivative calculation for EKF3 has been added here: https://github.com/priseborough/ardupilot/tree/pr-12439

See commit https://github.com/priseborough/ardupilot/commit/0c705b59fe29857aab652331645881f9640e476e

priseborough commented:
Increasing the value of EK2/3_HRT_FILT will make the complementary filter correct the IMU calculation of vertical velocity more aggressively using the baro data. A useful way to conceptualise this parameter is that below that frequency, height rate is predominantly calculated from the Baro data and above it is increasingly calculated from IMU data. At the cross-over frequency it will be a equal blend of each.

With noisy clipping affected IMU data, there will still be large high frequency noise in the height rate - that is unavoidable. We could add a first order LPF to the output if this is a problem, but this will be limited by the phase margin of the hgiht control loops.

rmackay9 commented:
@priseborough, great, thanks!  I've pulled it in and I'll test it tomorrow

rmackay9 commented:
@priseborough, thanks again for your help with this.  During testing I've found that if I set EK2_HRT_FILT to 100 a Floating point exception occurs.

Also I wanted to check on the naming of the new parameter.  I guess "HRT" doesn't refer to Height which we would normally name "HGT".  I guess it's perhaps **H**eight **R**eference **T**imeconstant?

rmackay9 commented:
@lthall and I completed the altitude control portion of this PR today.  The following steps need to be done before we can push this to master:

- re-test on a real vehicle
- resolve the floating point exception when the EKx_HGT_FILT is set to 40 or higher

khancyr commented on ArduCopter/ekf_check.cpp:
nitpick : this one isn't needed as it will be done L255

rmackay9 replied:
txs, removed in follow-up commit.

khancyr commented on ArduCopter/events.cpp:
```diff
@@ -6,6 +6,9 @@
 void Copter::failsafe_radio_on_event()
 {
+    // turn off artificial clipping on rc failsafe
+    ins.set_artificial_clipping(0);
```
I am pretty sure we let the drone in trouble with other messing-up the controler function when we get a failsafe, but still nice to get this one !

rmackay9 replied:
This particular change isn't going to go into master.  It's in one of the commits with "-- do not merge" in the title.  I'll remove it in the near future once we've completed all testing.

khancyr commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -595,7 +599,21 @@ void AC_PosControl::run_z_controller()
+        // during vibration compensation use feed forward with manually calculated gain
+        // ToDo: clear pid_info P, I and D terms for logging
+        if (!(_motors.limit.throttle_lower || _motors.limit.throttle_upper) || ((is_positive(_pid_accel_z.get_i()) && is_negative(_vel_error.z)) || (is_negative(_pid_accel_z.get_i()) && is_positive(_vel_error.z)))) {
+            _pid_accel_z.set_integrator(_pid_accel_z.get_i() + _dt * thr_per_accelz_cmss * 1000.0f * _vel_error.z * _p_vel_z.kP() * POSCONTROL_VIBE_COMP_I_GAIN);
```
shouldn't this be i - error * kp ?

lthall replied:
error is _vel_target.z - curr_vel.z so it should be i + error * kp

khancyr commented on libraries/AP_InertialNav/AP_InertialNav_NavEKF.cpp:
```diff
@@ -30,6 +31,14 @@ void AP_InertialNav_NavEKF::update()
     Vector3f velNED;
     if (_ahrs_ekf.get_velocity_NED(velNED)) {
+        // during high vibration events use vertical position change
+        // ToDo: fix and use EKF's get_vert_pos_rate() function instead of baro
```
todo fixed?

rmackay9 replied:
yes, txs.  comment removed in follow-up commit.

khancyr commented on libraries/AP_InertialSensor/AP_InertialSensor_SITL.cpp:
```diff
@@ -76,13 +76,13 @@ void AP_InertialSensor_SITL::generate_accel(uint8_t instance)
     if (vibe_freq.is_zero()) {
-        xAccel += accel_noise * rand_float();
-        yAccel += accel_noise * rand_float();
+        //xAccel += accel_noise * rand_float();
```
why remove ?

rmackay9 replied:
This is another change that won't be merged.  In SITL the vibrations are applied equally on X, Y and Z axis but this is not what we see on real vehicle (normally Z is much higher than X and Y).

This PR's change is to add protection for high Z-axis vibration so we remove the X and Y vibes because they interfere with testing.

khancyr commented on libraries/AP_NavEKF2/AP_NavEKF2_core.cpp:
```diff
@@ -751,6 +750,24 @@ void NavEKF2_core::calcOutputStates()
+    // Perform filter calculation using backwards Euler integration
+    // Coefficients selected to place all three filter poles at omega
+    const float CompFiltOmega = M_2PI * constrain_float(frontend->_hrt_filt_freq, 0.1f, 100.0f);
+    float omega2 = CompFiltOmega * CompFiltOmega;
```
nitpick const all !

rmackay9 commented:
I've test flown this on a real vehicle now with a patch to allow forcing the vibration failsafe on.  I tested both EKF2 and EKF3 and it performed quite well.
